p2pserver/dht/dht_message.go

```diff
@@ -0,0 +1,283 @@
+	log.Debugf("[dht]neighborsHandle: from %v ", from)
+}
+
+// pingHandle handles a ping message from UDP network
+func (this *DHT) pingHandle(from *net.UDPAddr, msg mt.Message) {
+	ping, ok := msg.(*mt.DHTPing)
+	if !ok {
+		log.Error("[dht]ping handle detected error message type!")
+		return
+	}
+	if ping.Version != this.version {
+		log.Errorf("[dht]pingHandle: version is incompatible. local %d remote %d",
+			this.version, ping.Version)
+		return
+	}
+
+	if ping.DestEndPoint.UDPPort != this.udpPort {
+		log.Errorf("[dht]pingHandle: udp port mismatch. input %d expected %d",
+			ping.DestEndPoint.UDPPort, this.udpPort)
+		return
+	}
+
+	// add the node to routing table
+	var node *types.Node
+	if node = this.routingTable.queryNode(ping.FromID); node == nil {
+		node = &types.Node{
+			ID:      ping.FromID,
+			IP:      from.IP.String(),
+			UDPPort: uint16(ping.SrcEndPoint.UDPPort),
+			TCPPort: uint16(ping.SrcEndPoint.TCPPort),
+		}
+	}
+	this.addNode(node)
+	this.pong(from)
+	log.Debugf("[dht]pingHandle: from %v ", from)
+}
+
+// pongHandle handles a pong message from UDP network
+func (this *DHT) pongHandle(from *net.UDPAddr, msg mt.Message) {
+	pong, ok := msg.(*mt.DHTPong)
+	if !ok {
+		log.Error("[dht]pong handle detected error message type!")
+		return
+	}
+	if pong.Version != this.version {
+		log.Errorf("[dht]pongHandle: version is incompatible. local %d remote %d",
+			this.version, pong.Version)
+		return
+	}
+
+	requestId := types.ConstructRequestId(pong.FromID, types.DHT_PING_REQUEST)
+	node, ok := this.messagePool.GetRequestData(requestId)
+	if !ok {
+		// request pool doesn't contain the node, ping timeout
+		log.Debugf("[dht]pongHandle: from %v timeout", from)
+		this.routingTable.removeNode(pong.FromID)
+		return
+	}
+
+	if pong.SrcEndPoint.UDPPort != node.UDPPort ||
+		pong.SrcEndPoint.TCPPort != node.TCPPort {
+		log.Errorf("[dht]pongHandle: mismatch port: input: udp %d tcp %d, expected: udp %d tcp %d",
+			pong.SrcEndPoint.UDPPort, pong.SrcEndPoint.TCPPort,
+			node.UDPPort, node.TCPPort)
+		return
+	}
+
+	// add to routing table
+	this.addNode(node)
+	// remove node from request pool
+	this.messagePool.DeleteRequest(requestId)
+	log.Debugf("[dht]pongHandle: from %v ", from)
+}
+
+// update the node to bucket when receive message from the node
+func (this *DHT) updateNode(fromId types.NodeID) {
+	node := this.routingTable.queryNode(fromId)
+	if node != nil {
+		// add node to bucket
+		bucketIndex, _ := this.routingTable.locateBucket(fromId)
+		this.routingTable.addNode(node, bucketIndex)
+	}
+}
+
+// findNode sends findNode to remote node to get the closest nodes to target
+func (this *DHT) findNode(remotePeer *types.Node, targetID types.NodeID) error {
+	addr, err := getNodeUDPAddr(remotePeer)
+	if err != nil {
+		return err
+	}
+	findNodeMsg := msgpack.NewFindNode(this.nodeID, targetID)
+	sink := comm.NewZeroCopySink(nil)
+	mt.WriteMessage(sink, findNodeMsg)
+	this.send(addr, sink.Bytes())
```

Comment:
to avoid duplicated serialization code , define `send` as:
```golang
func (this *DHT) send(addr *net.UDPAddr, msg Message) error {
       sink := comm.NewZeroCopySink(nil)
	mt.WriteMessage(sink, msg)
       ...
}
```